worker/src/direct_invocation/ws_direct_server_runner.rs

```diff
@@ -0,0 +1,108 @@
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+*/
+
+use crate::direct_invocation::ws_handler::CreateWsHandler;
+use log::*;
+use sgx_types::{sgx_enclave_id_t, sgx_status_t};
+use std::sync::Arc;
+use std::thread;
+use std::thread::JoinHandle;
+use ws::{listen, Handler};
+
+// TODO replace these extern C e-calls with a EnclaveAPI member field in the server impl
+extern "C" {
+    fn initialize_pool(eid: sgx_enclave_id_t, retval: *mut sgx_status_t) -> sgx_status_t;
+}
+
+/// Trait for a WebSocket Server for direct invocation
+pub trait RunWsServer {
+    fn run(&self, addr: String);
+}
+
+pub struct WsDirectServerRunner<H, HF>
+where
+    H: Handler,
+    HF: CreateWsHandler<Handler = H> + Sync + Send + 'static,
+{
+    handler_factory: Arc<HF>,
+    enclave_id: sgx_enclave_id_t,
+}
+
+impl<H, HF> RunWsServer for WsDirectServerRunner<H, HF>
+where
+    H: Handler,
+    HF: CreateWsHandler<Handler = H> + Sync + Send + 'static,
+{
+    fn run(&self, addr: String) {
+        let init_top_pool_handle = self.init_top_pool();
+
+        self.spawn_handler_thread(addr);
+
+        // ensure top pool is initialised before returning
+        init_top_pool_handle.join().unwrap();
+        println!("Successfully initialised top pool");
+    }
+}
+
+impl<H, HF> WsDirectServerRunner<H, HF>
+where
+    H: Handler,
+    HF: CreateWsHandler<Handler = H> + Sync + Send + 'static,
+{
+    pub fn new(handler_factory: Arc<HF>, enclave_id: sgx_enclave_id_t) -> Self {
+        WsDirectServerRunner {
+            handler_factory,
+            enclave_id,
+        }
+    }
+
+    fn init_top_pool(&self) -> JoinHandle<()> {
+        // initialise top pool in enclave
+        let enclave_id = self.enclave_id;
+
+        thread::spawn(move || {
+            let mut retval = sgx_status_t::SGX_SUCCESS;
+            let result = unsafe { initialize_pool(enclave_id, &mut retval) };
+
+            match result {
+                sgx_status_t::SGX_SUCCESS => {
+                    debug!("[TX-pool init] ECALL success!");
+                }
+                _ => {
+                    error!("[TX-pool init] ECALL Enclave Failed {}!", result.as_str());
+                }
+            }
+        })
+    }
+
+    fn spawn_handler_thread(&self, addr: String) {
+        info!("Starting direct invocation WebSocket server on {}", addr);
+        let handler_factory = self.handler_factory.clone();
+
+        thread::spawn(move || {
+            match listen(addr.clone(), |out| handler_factory.create(out)) {
+                Ok(_) => (),
+                Err(e) => {
+                    error!(
+                        "error starting worker direct invocation api server on {}: {}",
+                        addr, e
+                    );
+                }
+            };
+        });
```

Comment:
uses the (abstract) handler factory to create a WS handler instance. This was required, because the `listen` method takes a factory method to construct a handler by supplying a `Sender` (called `out` here)

Reply:
In my opinion, this should actually be the pattern when defining functions, as opposed to using a factory trait as arg.

`|| -> MySomething` can also be used with our custom factories, i.e., `|| -> SomethingFactory.create()`.

Reply:
It's essentially the same. The advantage of having a factory trait, is that you can name the trait and make it obvious what it does. Having the argument as function can be a bit more difficult to read in my opinion, especially when the function gets more complex with multiple arguments (which can be the case for factories). You will end up with something like:
```
fn do_this<MyType>(other_arg: A, factory: Fn(i32, DepA, OtherThing) -> MyType) {
    let my_thing = factory(a, b, c);
}
```

as opposed to:

```
fn do_this<MyFactory>(other_arg: A, factory: MyFactory) {
    let my_thing = factory.create(a,b,c);
}
```

Reply:
Well you could also do this: 

```
fn do_this<MyTypeConstructor>(other_arg: A, factory: MyTypeConstructor) 
where:
     MyTypeConstructor: Fn(String, SomeOtherType) -> MyType
{
    let my_thing = factory.create(a,b);
}
```

However, I agree that we face a trade-off between readability and flexibility.